pkg/cni/types.go

```diff
@@ -0,0 +1,73 @@
+package cni
+
+import (
+	"context"
+	"github.com/containernetworking/cni/pkg/invoke"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/k8sclient"
+	"gopkg.in/k8snetworkplumbingwg/multus-cni.v3/pkg/types"
+	"net/http"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/types/current"
+)
+
+type cniRequestFunc func(request *PodRequest) ([]byte, error)
+
+// Explicit type for CNI commands the server handles
+type command string
+
+// PodRequest represents a request for ADD / DEL / CHECK for a Pod
+type PodRequest struct {
+	// The CNI command of the operation
+	Command command
+	// kubernetes namespace name
+	Namespace string
+	// kubernetes pod name
+	Name string
+	// kubernetes pod UID
+	UID string
+	// kubernetes container ID
+	SandboxID string
+	// CNI container ID
+	ContainerID string
+	// kernel network namespace path
+	Netns string
+	// Interface name to be configured
+	IfName string
+	// CNI conf obtained from stdin conf
+	CNIConf *types.NetConf
+	// Timestamp when the request was started
+	timestamp time.Time
+	// ctx is a context tracking this request's lifetime
+	ctx context.Context
+	// cancel should be called to cancel this request
+	cancel context.CancelFunc
+	// kubeclient has the kubernetes API client
+	kubeclient *k8sclient.ClientInfo
+	// exec execs
+	exec invoke.Exec
+}
```

Comment:
Instead of relying on a payload that fits libcni - `skel.CmdArgs` -  I am adding one that fits multus, and have added this mix of runtime config (`ContainerID`, `NetNS`, `IfName`), `K8sArgs` (pod name, pod namespace, pod UID, and pod sandbox ID), utilities required for handling a request - kubeclient, exec, etc. My intention was to gather together all the required data to handle a request to configure / cleanup networking for a pod.

I don't want to handle CmdArgs [StdinData](https://github.com/k8snetworkplumbingwg/multus-cni/blob/5feb1343cd8318a578d2c06b8c678450e7217068/vendor/github.com/containernetworking/cni/pkg/skel/skel.go#L42) nor `args`. Nor want to use `types.UnmarshableStrings` unless I really have to.

I did not understand if your issue is with me introducing those two types (`Cmd` and `PodRequest`) or if you want both of them gone.

Reply:
Regarding Stdin, it should be parsed and it is in https://github.com/k8snetworkplumbingwg/multus-cni/blob/master/pkg/types/types.go#L27-L59 so you don't need to handle raw stdin as long as you use `types.LoadNetConf()`. 

Gathering the required info as structure may make sense to me, but I don't know why you need two structure for that.
To gathering info, one structure seems to be enough, at least to me. But you introduce two structures.
As you mentioned above, your purpose to introduce them is 'My intention was to gather together all the required data to handle a request to configure / cleanup networking for a pod.', but your reason is not enough to me to explain 'why you introduce two structure'.

I suppose you have more reason why you need two structure for one purpose. So first, I would like to clarify your reason.
Please let me know why you need both.

Reply:
`Cmd` holds the runtime data (part of libcni's runtime config (ContainerID, NetNS, IfName) plus multus `K8sArgs`, holding the relevant runtime cni arguments (pod name / uid / namespace + sandbox ID). I am forced to agree its name could be better.

`PodRequest` holds that, plus the multus configuration, kubeclient, request timestamp (logging purposes), request context + cancel function, and cni exec caller.

To me, these are different structure, with different goals and are used in different scopes. I can embed one within the other; imo, it would make the relationship between them more explicit. 

Your thoughts @s1061123 ?

Reply:
If you design code from scratch, the structure may make sense, however,
currently pkg/multus already have CmdAdd, CmdDel and CmdCheck based on
libcni's API structure, skel.CmdArgs. In CmdAdd/CmdDel/CmdCheck, NetConf
structure, DelegateNetConf structure and other structures are used to collect
required information, such as Kubernetes pod info and so on.

Once you substitute your request to skel.CmdArgs and call
CmdAdd/CmdDel/CmdCheck, current multus code does all processing for you.
You don't need to handle stdin data and NetConf because current code handles
for you because current multus has the code without any new structure
introduction.

Your PR seems to change the current code a lot to introduce the new structure,
not utilizing current structure. To me, I don't know why you need these
structures. Please let me know why you need these structures, because current
multus does these process without your structures.

In addition, I still need your answer for why the strucutre is introduced
because some variables still ambiguous use, like 'timestamp time.Time' and
'exec invoke.Exec' in 'PodRequest'. Both variable is set once and never used
(refer/update) as far as I looked your code however, I believe that you
have clear usecase for that and it must be used in your PR with clear
purposes, because you added these variables in the PR.

Reply:
...

> Once you substitute your request to skel.CmdArgs and call CmdAdd/CmdDel/CmdCheck, current multus code does all processing for you. You don't need to handle stdin data and NetConf because current code handles for you because current multus has the code without any new structure introduction.

While I'm aware of that advantage, modelling the internal communication of the multus processes using a libcni API structure does not seem a good choice (to me). I do not think the multus shim / daemon should speak CNI between themselves.

> 
> Your PR seems to change the current code a lot to introduce the new structure, not utilizing current structure. To me, I don't know why you need these structures. Please let me know why you need these structures, because current multus does these process without your structures.

I'm proposing to keep the current structure, and use an intermediate canonical representation, which both the thin and thick options use: they both speak CNI in its API (both the multus-shim and multus thin plugin) and **uses** this intermediate representation as an implementation detail.

> 
> In addition, I still need your answer for why the strucutre is introduced because some variables still ambiguous use, like 'timestamp time.Time' and 'exec invoke.Exec' in 'PodRequest'. Both variable is set once and never used (refer/update) as far as I looked your code however, I believe that you have clear usecase for that and it must be used in your PR with clear purposes, because you added these variables in the PR.

`exec` is used (and passed to) libcni, check the following links:
  - https://github.com/k8snetworkplumbingwg/multus-cni/pull/771/files#diff-ef8a5ed2f4065ae76aaaefd57d7cacf3956c1c0a217fe8a45bd75b74197e858dR257
  - https://github.com/k8snetworkplumbingwg/multus-cni/pull/771/files#diff-ef8a5ed2f4065ae76aaaefd57d7cacf3956c1c0a217fe8a45bd75b74197e858dR281
  - https://github.com/k8snetworkplumbingwg/multus-cni/pull/771/files#diff-ef8a5ed2f4065ae76aaaefd57d7cacf3956c1c0a217fe8a45bd75b74197e858dR301

I forgot to log how long does it take to process a request - that's the purpose of the timestamp attribute. I'll add that asap; thanks for spotting it.

Are there any more attributes you consider ambiguous ?

Reply:
> While I'm aware of that advantage, modelling the internal communication of the multus processes using a libcni API structure does not seem a good choice (to me). I do not think the multus shim / daemon should speak CNI between themselves.

I don't know why. Especially you design that the multus shim / daemon speak CNI between themselves, even though you told that 'I do not think the multus shim / daemon should speak CNI between themselves.' If you think the multus shim / daemon should *NOT* speak CNI between themselves, then you just design and write a code for that.

I would like to know why you feel that modelling the internal communication of the multus processes using a libcni API structure does not seem a good choice.
Please let me know clear and objective reason why.

> exec is used (and passed to) libcni, check the following links:

I am asking that why you add exec variable in PodRequest, not why you use exec.
As you mentioned and wrote in a comment, PodRequest contains per Pod basis request, which means PodRequest should have pod related information.

Looking the 'exec' variable in PodRequest, it seems to following:

1. set the variable at [cniRequestToPodRequest()](https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L150), which is come from [newCNIServer()](https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L97), that is nil as https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L85
2. then this goes to cmdAdd()/cmdCheck()/cmdDel(), without any update.

Hence, the exec variable always initial variable from server, that is nil.
I'm wondering why you set nil in the variable and hold it.

As far as I know of libcni code, exec does not contain any information related to per Pod basis information however you put this variable as per Pod basis information. I would like to know why you add because you may have some special reason.

During to write reply, you already changed the PR. So I will check your code again and reply it.

Reply:
> I don't know why. Especially you design that the multus shim / daemon speak CNI between themselves, even though you told that 'I do not think the multus shim / daemon should speak CNI between themselves.' If you think the multus shim / daemon should _NOT_ speak CNI between themselves, then you just design and write a code for that.

Hm, I thought I was doing that. They're speaking `PodRequest` / `PodResponse` - which **do** encapsulate the libcni's structs.

> 
> I would like to know why you feel that modelling the internal communication of the multus processes using a libcni API structure does not seem a good choice. Please let me know clear and objective reason why.

Being a CNI plugin, multus must understand CNI. As I see it, **at the edges** - it must be invoked via CNI API, and must in turn invoke another plugin, via CNI API. 

I don't think modelling the internal communication of 2 processes using libcni's structs is a good choice because that is exposing an implementation detail.

Furthermore, I'd like to remind you that I'm using pod request to group together all I need to handle a request - and not everything is CNI related - I need to inject the kubeclient, have a timestamp, exec (keep reading for the explanation on this ), etc. The multus daemon will eventually invoke the multus library funcs (`CmdAdd` / ... ), and for that, it'll need to pass kubeclient and exec - which, I need to mock in the tests. 

I've embedded the libcni / k8sArgs into the `PodRequest`. Is that good enough for you ? 

> 
> > exec is used (and passed to) libcni, check the following links:

> 
> I am asking that why you add exec variable in PodRequest, not why you use exec. As you mentioned and wrote in a comment, PodRequest contains per Pod basis request, which means PodRequest should have pod related information.

I miss a few logic steps on the claim that "PodRequest should have pod related information". My goal is to have everything together to **handle a request** ...  That includes the arguments required for `multus.CmdAdd` / `multus.CmdDel` / `multus.CmdCheck` - which are also kubeclient and exec. I'd also like to have a timestamp, so I can figure out how much time we spend on a request.

Could it be I should rename the struct ? 

> 
> Looking the 'exec' variable in PodRequest, it seems to following:
> 
> 1. set the variable at [cniRequestToPodRequest()](https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L150), which is come from [newCNIServer()](https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L97), that is nil as https://github.com/maiqueb/multus-cni/blob/thick-arch/pkg/cni/server.go#L85
> 2. then this goes to cmdAdd()/cmdCheck()/cmdDel(), without any update.
> 
> Hence, the exec variable always initial variable from server, that is nil. I'm wondering why you set nil in the variable and hold it.

On the production code, you're right. But on the tests I pass a `fakeExec` to the server, which allows me to mock the exec handler and unit test the server / client communication. 

https://github.com/k8snetworkplumbingwg/multus-cni/pull/771/files#diff-8998a16bc50da50c38ab8158dfb00d4a307f3dd01f4d7ec708e0005a503b18d3R210

> 
> As far as I know of libcni code, exec does not contain any information related to per Pod basis information however you put this variable as per Pod basis information. I would like to know why you add because you may have some special reason.

I just need a way to control what exec is used on the requests - as such I inject the exec from the server into every request. 

> 
> During to write reply, you already changed the PR. So I will check your code again and reply it.

Reply:
Removed `PodRequest`. Thus, I think this conversation can be resolved.

Let me know if I'm wrong.